src/tpm.rs

```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
+        let v: Vec<&str> = ret_out.matches(&curr_handle).collect();
+        if !v.is_empty() {
+            run(
+                format!(
+                    "tpm2_evictcontrol -a o -c {} -P {}",
+                    hex::encode(curr_handle),
+                    owner_pw
+                ),
+                None,
+            )?;
+        }
+    }
+
+    if owner_pw.is_empty() {
+        owner_pw = random_password(20)?;
+        set_tpm_metadata_content("owner_pw", &owner_pw);
+    }
+
+    let ek_pw = random_password(20)?;
+    let tf = NamedTempFile::new()?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let paths: Vec<&str> = vec![tf_path];
+    let (ret_out, f_out) = run(
+        format!(
+            "tpm2_createek -c -G {} -p {} -P {} -o {} -e {}",
+            ek_asym_alg, tf_path, ek_pw, owner_pw, owner_pw
+        ),
+        Some(paths),
+    )?;
+
+    let ek_tpm = f_out.get(tf_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "Createek output to file content is missing",
+        )
+    })?;
+
+    let ret_out_map: Value = serde_yaml::from_str(&ret_out)?;
+    let persistent_handle = ret_out_map["persistent-handle"]
+        .as_str()
+        .unwrap_or_else(|| "");
+
+    set_tpm_metadata_content("ek_handle", persistent_handle.into())?;
+    set_tpm_metadata_content("ek_pw", &ek_pw);
+    set_tpm_metadata_content("ek_tpm", &base64::encode(&ek_tpm));
+    Ok(())
+}
+
+/*
+ *  Return:
+ *      Failed if retrieve public key process failed.
+ *
+ *  Retrieve the public key of ek key and store it in the tpmedata.json file.
+ */
+fn tpm_get_pub_ek() -> Result<(), KeylimeTpmError> {
+    let ek_handle = get_tpm_metadata_content("ek_handle")?;
+    let tf = NamedTempFile::new()?;
+    let tf_path = temp_file_get_path(&tf)?;
+    let paths = vec![tf_path];
+
+    let (_, ret_file_map) = run(
+        format!(
+            "tpm2_readpublic -c {} -o {} -f pem",
+            hex::encode(ek_handle),
+            tf_path
+        ),
+        Some(paths),
+    )?;
+
+    let ek = ret_file_map.get(tf_path).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "tpm2_readpublic fail, ek key is missing is output file.",
+        )
+    })?;
+    set_tpm_metadata_content("ek", &ek);
+    Ok(())
+}
+
+/*
+ *  Return;
+ *      Error if the flush process failed
+ *
+ *  Retrieve all the persistent object and remove those object.
+ */
+fn flush_keys() -> Result<(), KeylimeTpmError> {
+    let (ret_out, _) = run("tpm2_getcap -c handles-persistent".into(), None)?;
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let ret_out_val: Value = serde_yaml::from_str(&ret_out)?;
+    let ret_out_map: &Map<String, Value> =
+        ret_out_val.as_object().ok_or_else(|| {
+            KeylimeTpmError::new_tpm_rust_error("Output invalid.")
+        })?;
+    ret_out_map.iter().for_each(|(k, _)| {
+        run(
+            format!(
+                "tpm2_evictcontrol -a o -c {} -P {}",
```

Comment:
you guessed it :)
https://github.com/keylime/keylime/blob/master/keylime/tpm2.py#L515